sysdump/sysdump.go

```diff
@@ -1448,6 +1452,53 @@ func (c *Collector) Run() error {
 		return fmt.Errorf("failed to close the worker pool: %w", err)
 	}
 
+	// Run each serial task in its own workerpool.
+	for i, t := range serialTasks {
+		t := t
+		if c.shouldSkipTask(t) {
+			c.logDebug("Skipping %q", t.Description)
+			continue
+		}
+
+		var wg sync.WaitGroup
+		if t.CreatesSubtasks {
+			wg.Add(1)
+		}
+
+		// Adjust the worker count to make enough headroom for tasks that submit sub-tasks.
+		// This is necessary because 'Submit' is blocking.
+		wc := 1
+		if t.CreatesSubtasks {
+			wc++
+		}
+		c.Pool = workerpool.New(wc)
+
+		// Add the serial task to the worker pool.
+		if err := c.Pool.Submit(fmt.Sprintf("[%d] %s", len(tasks)+i, t.Description), func(ctx context.Context) error {
+			if t.CreatesSubtasks {
+				defer wg.Done()
+			}
+			c.logTask(t.Description)
+			defer c.logDebug("Finished %q", t.Description)
+			return t.Task(ctx)
+		}); err != nil {
+			return fmt.Errorf("failed to submit task to the worker pool: %w", err)
+		}
+
+		// Wait for the subtasks to be submitted and then call 'Drain' to wait for them to finish.
+		wg.Wait()
```

Comment:
It's a bit sad that the worker pool requires a separate WG in order to operate properly. Can't we treat submissions to the Pool as blocking?

I see that this is the pattern that we use elsewhere. Seems fine.

Reply:
I'm afraid we need this because of the subtasks. Each task in `serialTasks` is executed asynchronously, and they too might submit a new subtask to the same pool. (In fact, we are doing this in the `gops trace` case, inserting a subtask for each agent to trace).
But calling `Drain` prevents subtasks to do that:

```
// Drain waits until all tasks are completed. This operation prevents
// submitting new tasks to the worker pool. Drain returns the results of the
// tasks that have been processed.
```

So, without the wg this could happen:

```
1) submit main task
2) call Drain -> wait for main task to finish
3) execute the task
  3a) `c.Pool.Submit` in `SubmitTracingGopsSubtask` is gonna fail because `wp` is in draining mode
```

I agree with you that this is a bit cumbersome, probably we could probably use another method in the workerpool to wait for tasks termination without entering draining mode.
I see that the same pattern has been used for previous `tasks` too.

Reply:
Interleaving the calls to `Submit` seems like a similarly poor API to me. I agree that this is the established pattern and it's certainly fine for now.